crates/fuel-core/src/state/rocks_db.rs

```diff
@@ -499,6 +505,55 @@ where
         }
     }
 
+    fn reverse_prefix_iter_keys(
```

Comment:
The `reverse_prefix_iter` function is pretty complex on its own, and having two complex functions will be hard to support=D Maybe we could use the generic, which will define what type to return and defines the closure to get that type?

Reply:
totally agree, however, since the types we want to implement for, namely `KVItem` and `KeyItem` are a tuple and a distinct type, writing a generic for this becomes more involved. We will have to define a trait for both `KVItem` and `KeyItem` which provide the following helper fn -

```rust
trait StorageItem {
  type Key = Vec<u8>;

  fn key() -> Self::Key,
  fn key_starts_with(prefix: Self::Key) -> bool; 
}
```

which makes it easier to write the function, in the following way -


```rust
fn reverse_prefix_iter_generic<T, F>(
        &self,
        prefix: &[u8],
        column: Description::Column,
        iter_store_fn: F,
        iter_all_fn: impl Fn(
            &Self,
            Description::Column,
            ReadOptions,
            OwnedIteratorMode,
        ) -> BoxedIter<StorageResult<T>>,
    ) -> BoxedIter<StorageResult<T>>
    where
        T: 'static,
        F: Fn(
            &Self,
            Description::Column,
            Option<&[u8]>,
            Option<&[u8]>,
            IterDirection,
        ) -> BoxedIter<'static, StorageResult<T>>,
    {
        let maybe_next_item = next_prefix(prefix.to_vec())
            .and_then(|next_prefix| {
                iter_store_fn(
                    self,
                    column,
                    Some(next_prefix.as_slice()),
                    None,
                    IterDirection::Forward,
                )
                .next()
            })
            .and_then(|res| res.ok());

        let prefix_vec = prefix.to_vec(); 
        
        match maybe_next_item {
            Some(key_or_item) => {   
                let next_start_key_vec = key_or_item.key().to_vec();

                let iter_mode = OwnedIteratorMode::From(
                    next_start_key_vec.to_vec(),
                    rocksdb::Direction::Reverse,
                );

                iter_all_fn(self, column, self.read_options(), iter_mode)
                    .skip(1)
                    .take_while(move |item| match item {
                        Ok(key_or_item) => key_or_item.key_starts_with(&prefix_vec), 
                        Err(_) => true,
                    })
                    .into_boxed()
            }
            None => {
                iter_all_fn(self, column, self.read_options(), OwnedIteratorMode::End)
                    .take_while(move |item| match item {
                        Ok(key_or_item) => key_or_item.key_starts_with(&prefix_vec), 
                        Err(_) => true,
                    })
                    .into_boxed()
            }
        }
    }
```

the important line of code in the above fn being `let next_start_key_vec = key_or_item.key().to_vec();`

to avoid destructuring the tuple since the generic `T` can either be a KVItem (a tuple) or a KeyItem (Vec<u8>)

perhaps this can be addressed in a follow up PR to avoid a large diff :)

what are your thoughts?

Reply:
@xgreenx check out the diff in e44976669, I added it there and if we want to include as a follow up pr, i can create it